policy: improve policy computation for high cadinality of http rules.

marseel opened the pull request:
See commit message for more details.

High-level overview:
- Add two tests for a bugs where we don't remove duplicated PortRuleHTTP or rules with duplicated Headers/HeaderMatches
- Add benchmark for high cardinality of rules and/or PortRuleHTTP within a single rule
- Improve policy calculation by performing deduplication only once, additionally fixing a bug mentioned above
- Simplify sorting in XDS as we perform most of the sorting in policy engine now.

```release-note
policy: improved performance of policy calculation with high number of http rules
```

Before:
```
BenchmarkRegenerateL7PolicyRules-32     3936    rules/sec
BenchmarkRegenerateL7PolicyRules2-32    3800    rules/sec
BenchmarkRegenerateL7PolicyRules3-32    111554  rules/sec
BenchmarkRegenerateL7PolicyRules4-32    15609   rules/sec
```
After:
```
BenchmarkRegenerateL7PolicyRules-32     721067  rules/sec
BenchmarkRegenerateL7PolicyRules2-32    768899  rules/sec
BenchmarkRegenerateL7PolicyRules3-32    301971  rules/sec
BenchmarkRegenerateL7PolicyRules4-32    313559  rules/sec
```

marseel committed:
policy: improve policy computation for high cadinality of http rules.

Previously, merging all http rules had complexity of O((n*m)^2), where n is
number of rules in repository for given port and m is number of http rules.
This was problematic, especially when identity had tens of rules with
hundreds or thousands of http rules each.
As a result, pod selected by such policies was unable to start, as
policy computation did not finish within 60s timeout of CNI ADD.

New approach, which sorts http rules and removes duplicates once,
has O((n*m) log (n*m)), which is more reasonable.

Benchmark results:
BenchmarkRegenerateL7PolicyRules-32     721067  rules/sec
BenchmarkRegenerateL7PolicyRules2-32    768899  rules/sec
BenchmarkRegenerateL7PolicyRules3-32    301971  rules/sec
BenchmarkRegenerateL7PolicyRules4-32    313559  rules/sec

which is between 3-200 times improvement over previous implementation
depending on use-case.

Fixes a bug where duplicated PortRuleHTTPs within a
single rule were not deduplicated.

Signed-off-by: Marcel Zieba <marcel.zieba@isovalent.com>

marseel committed:
policy: add test for duplicated PortRuleHTTP

This test fails with current implementation as we blindly copy
PortRuleHTTP for the first rule skipping removal of duplicates.

Signed-off-by: Marcel Zieba <marcel.zieba@isovalent.com>

marseel committed:
policy: add benchmark for high cardinality of PortRuleHTTP

Add couple of benchmarks with high cardinality of rules and high
cardinality of PortRuleHTTP within a single rule.

Example results:
BenchmarkRegenerateL7PolicyRules-32     3936    rules/sec
BenchmarkRegenerateL7PolicyRules2-32    3800    rules/sec
BenchmarkRegenerateL7PolicyRules3-32    111554  rules/sec
BenchmarkRegenerateL7PolicyRules4-32    15609   rules/sec

Signed-off-by: Marcel Zieba <marcel.zieba@isovalent.com>

aanm commented on pkg/policy/api/utils.go:
```diff
@@ -17,42 +18,66 @@ func (h *PortRuleHTTP) Exists(rules L7Rules) bool {
+		return -1
 	}
-	return true
+	return cmp.Or(
+		cmp.Compare(h.Path, o.Path),
+		cmp.Compare(h.Method, o.Method),
+		cmp.Compare(h.Host, o.Host),
+		slices.Compare(h.Headers, o.Headers),
+		slices.CompareFunc(h.HeaderMatches, o.HeaderMatches, func(a, b *HeaderMatch) int {
+			return a.Compare(b)
+		}),
+	)
```
Is this fast than `==`? :exploding_head:

squeed commented on pkg/policy/api/utils.go:
```diff
@@ -17,42 +18,66 @@ func (h *PortRuleHTTP) Exists(rules L7Rules) bool {
+		cmp.Compare(h.Path, o.Path),
+		cmp.Compare(h.Method, o.Method),
+		cmp.Compare(h.Host, o.Host),
+		slices.Compare(h.Headers, o.Headers),
```
We should be order-agnostic, no?

marseel replied:
Interesting. Previous implementation seems to be incorrect as well in that case :thinking:

marseel replied:
Okay, I have an interesting idea. How about we keep it simple in policy engine and just append all http/kafka etc rules without sorting/deduplication and then we do sorting/deduplication here:
https://github.com/cilium/cilium/blob/024288e8308617f980adf1e602006b9568b3c1b0/pkg/envoy/policy/envoy_l7_rules_translator.go#L53
Especially as we already do headers sorting here: https://github.com/cilium/cilium/blob/024288e8308617f980adf1e602006b9568b3c1b0/pkg/envoy/policy/envoy_l7_rules_translator.go#L232

That seems like a more reasonable solution to me to split responsibilities between envoy/policy engine, WDYT?

marseel committed:
xds: simplify PortNetworkPolicySlice sorting

We were not taking into account EndPort when sorting portRules.
Additionally, simplify sorting as we don't need to sort based on rules
as (Port, EndPort, Protocol) is unique.

Signed-off-by: Marcel Zieba <marcel.zieba@isovalent.com>

marseel committed:
xds: simplify PortNetworkPolicyRuleLess sorting

Instead of sorting based on the content of rules, let's sort based on
selector key.

Signed-off-by: Marcel Zieba <marcel.zieba@isovalent.com>

marseel committed:
xds: stop sorting HTTP rules and rely on policy engine

Since we sort rules within policy engine, there is no need to perform
additional sorting in translation/xds layer.

Signed-off-by: Marcel Zieba <marcel.zieba@isovalent.com>

marseel committed:
l7/http: sort headers and headermatches within rules

Fixes a test with duplicated headers/headersMatches.

Signed-off-by: Marcel Zieba <marcel.zieba@isovalent.com>

marseel committed:
policy: add test for merging duplicated headers/headersMatch

This test currently fails as we are not sorting headers or
headersMatches before removing duplicates.

Signed-off-by: Marcel Zieba <marcel.zieba@isovalent.com>

marseel committed:
tests: fix printing policy diff

Previously, the output contained:
Messages:   	Missing (-), Unexpected (+):
Where Unexpected was expected result ("want") while Missing was actual
result.
Instead, it should be Missing == expected, Unexpected == result

Signed-off-by: Marcel Zieba <marcel.zieba@isovalent.com>
